src/Microsoft.DocAsCode.Metadata.ManagedReference/ExtractMetadataWorker.cs

```diff
@@ -20,50 +19,52 @@ namespace Microsoft.DocAsCode.Metadata.ManagedReference
-        private static readonly string[] SupportedProjectName = { "project.json" };
-        private static readonly string[] SupportedProjectExtensions = { ".csproj", ".vbproj" };
-        private static readonly string[] SupportedSourceFileExtensions = { ".cs", ".vb" };
-        private static readonly string[] SupportedVBSourceFileExtensions = { ".vb" };
-        private static readonly string[] SupportedCSSourceFileExtensions = { ".cs" };
-        private static readonly string[] SupportedAssemblyExtensions = { ".dll", ".exe" };
-        private static readonly string SupportedCommentFileExtension = ".xml";
-        private static readonly List<string> SupportedExtensions = new List<string>();
-        private readonly ExtractMetadataInputModel _validInput;
-        private readonly ExtractMetadataInputModel _rawInput;
+        private const string XmlCommentFileExtension = "xml";
+        private const string IndexFileName = ".manifest";
+        private readonly Dictionary<FileType, List<FileInfo>> _files;
         private readonly bool _rebuild;
         private readonly bool _shouldSkipMarkup;
         private readonly bool _preserveRawInlineComments;
-        private readonly string _filterConfigFile;
         private readonly bool _useCompatibilityFileName;
-
-        // TODO: refactor incremental logic
+        private readonly string _filterConfigFile;
+        private readonly string _outputFolder;
         private readonly Dictionary<string, string> _msbuildProperties;
+
+        //Lacks UT for shared workspace
         private readonly Lazy<MSBuildWorkspace> _workspace;
-        static ExtractMetadataWorker()
-        {
-            SupportedExtensions.AddRange(SupportedSolutionExtensions);
-            SupportedExtensions.AddRange(SupportedProjectExtensions);
-            SupportedExtensions.AddRange(SupportedSourceFileExtensions);
-            SupportedExtensions.AddRange(SupportedAssemblyExtensions);
-        }
 
-        public ExtractMetadataWorker(ExtractMetadataInputModel input, bool rebuild, bool useCompatibilityFileName)
+        public ExtractMetadataWorker(ExtractMetadataInputModel input)
         {
-            _rawInput = input;
-            _validInput = ValidateInput(input);
-            _rebuild = rebuild;
+            if (input == null)
+            {
+                throw new ArgumentNullException(nameof(input));
+            }
+
+            if (string.IsNullOrEmpty(input.OutputFolder))
+            {
+                throw new ArgumentNullException(nameof(input.OutputFolder), "Output folder must be specified");
+            }
+
+            if (input.Files == null || input.Files.Count == 0)
+            {
+                Logger.Log(LogLevel.Warning, "No source project or file to process, exiting...");
+                return;
+            }
+
+            _files = ExtractMetadataUtility.GetInputs(input.Files).GroupBy(f => f.Type).ToDictionary(s => s.Key, s => s.Distinct().ToList());
```

Comment:
yes, absolutely:thumbsup: 

---
In reply to: [118815303](https://github.com/dotnet/docfx/pull/1707#discussion_r118815303) [](ancestors = 118815303)